### ISSUE TITLE
expand-region does not work as expected

### ISSUE THREAD

**Issue by vadali** [magnars/expand-region.el#201]
Hey, 
Expand region recently started to mis-behave when trying to select symbols in a lisp language (specifically clojure and emacs-lisp). I tried to follow the steps given in #195, but uninstalling/reinstalling/recompiling did not have the desired effects.

Description of the problem:
If point is in the middle of a symbol, it will select the rest of the symbol from point forward on the first C-= invocation, then until the end of the s-exp, then the entire s-exp. I'd like the second invocation to select the symbol itself.

Any clues?

**Comment by vadali** [comment]
As a note, if I manually run `er/mark-symbol` it works just fine..

**Comment by vadali** [comment]
Looking further into it, I set `(setq er--show-expansion-message t)` and added `(message "Trying %S" (car try-list))` right after `(funcall (car try-list))` in `er--expand-region-1`. 

Here is a test case (| is point):
`
(sym1 sy|m2 sym3)
`

hitting `C-=` once prints this into `*Messages*`:
`
Trying er/mark-word
er/mark-word
Trying er/mark-symbol
Trying er/mark-symbol-with-prefix
Trying er/mark-next-accessor
Trying er/mark-method-call
Trying er/mark-inside-quotes
er/mark-inside-quotes
Trying er/mark-outside-quotes
Trying er/mark-inside-pairs
Trying er/mark-outside-pairs
Trying er/mark-comment
Trying er/mark-defun
Type = to expand again, - to contract, 0 to reset
`

This shows that `er/mark-symbol` is running, however `er--this-expansion-is-better` returns false for it..

**Comment by magnars** [comment]
This is a known issue that is normally fixed by reinstalling, so that is pretty weird. Your emacs version?

**Comment by vadali** [comment]
hmm, its actually head (emacs-25.1-rc2-126555-ge951aa1), interesting

**Comment by magnars** [comment]
I generally don't support non-stable versions, since I've seen so much breakage through the years. I'll dive into it if the problem persists after an official release.

**Comment by vadali** [comment]
That makes sense.. I also checked it on 24.5.1 and it seemed to work fine. The version I have installed is 0.10.0 (melpa-stable). I also tested 20160729.131 and I get the same behavior.

For what its worth, tested on an .el file - it works well if the symbol is not wrapped in parens, otherwise I get this behavior. 

Could you perhaps point me where to look? I started looking into `er--expand-region-1` and `er--this-expansion-is-better` but felt that I was barking up the wrong tree or at least branch.

I use(d) this package quite heavily and really don't mind putting some efforts in making it work again.

Thanks!

**Comment by vadali** [comment]
Tried to systematically remove functions from `er/try-expand-list`, and removing `er/mark-comment` seemed to make things working again.. 

However, this is really strange - when debugging the function, it only calls `er--point-is-in-comment-p` and exists (because the predicate returns nil). I deleted all lines from the function `mark-comment` and the problem happened again. I even tried to make my own empty function `(defun test-func ())` and it also causes the same problematic behavior.. why would this happen? Is there some "stateful contract" that has to be followed by potential mark functions? 

In any case, removing this function makes it work for now.

**Comment by magnars** [comment]
Wow, that is reallly weird. I can't think of why that happens. Thanks for
investigating tho.

**Comment by vadali** [comment]
ok. cant explain this, but it randomly started working again - with `er/mark-comment` :| 

I dont know how come - I completely removed `.emacs.d/elpa` directory before every test, get all my packages re-downloaded and re-compiled, and for some reason it just started to work again...

voodoo in action.

**Comment by magnars** [comment]
That is decidedly weird.

**Comment by photex** [comment]
With the official release of Emacs 25.1 I can confirm this is an issue still.

I haven't tried removing and reinstalling anything yet. I'm using OSX and Windows, the behavior is the same as described on both platforms.

**Comment by magnars** [comment]
Yeah, the way expand-region is compiled means you'll have to clean out old .elc-files and restart/recompile to get it up again.

**Comment by mkleehammer** [comment]
I'm still seeing this on 25.1 for Windows after recompiling all .el files.

It is 0.10 from MELPA-Stable, though.

**Comment by magnars** [comment]
You need to clean out all .elc-files, and also restart Emacs.

**Comment by Trevoke** [comment]
I get strange broken behavior as well on emacs 25, on OSX (definitely compiled for emacs 25)

For Scala (expand-region twice):

![image](https://cloud.githubusercontent.com/assets/67913/19241496/e8b27532-8edc-11e6-8484-a9484fec7ff8.png)

For elisp (expand-region once):

![image](https://cloud.githubusercontent.com/assets/67913/19241526/0ba1a91e-8edd-11e6-9bc4-70e5a0c10333.png)


Javascript (expand-region once, js2-mode):

![image](https://cloud.githubusercontent.com/assets/67913/19241562/30d2e3a6-8edd-11e6-956a-2d7260c8ce16.png)


Behavior seems to work well with Ruby (enh-ruby-mode).

Elixir (expand-region once):

![image](https://cloud.githubusercontent.com/assets/67913/19241634/7e8a3ce8-8edd-11e6-86ed-b91164142f54.png)

**Comment by Trevoke** [comment]
Okay, this is interesting.. Check out the difference in behavior for expand-region, depending on where the point is:

![image](https://cloud.githubusercontent.com/assets/67913/19241767/2f465d5a-8ede-11e6-813b-939fb029b3f1.png)

![image](https://cloud.githubusercontent.com/assets/67913/19241773/359c3f44-8ede-11e6-8b91-8e429aef6576.png)

**Comment by dsedivec** [comment]
> I get strange broken behavior as well on emacs 25, on OSX (definitely compiled for emacs 25)

FWIW I was having this problem (Emacs 25 on Mac) until I recompiled *all* of my third-party Emacs packages under Emacs 25, not just expand-region.  I use el-get so I blew away `~/.emacs.d/el-get` and just reinstalled everything.  The problem has not recurred since then, and that was probably at least two weeks ago.

(Of course I expect it wasn't necessary to recompile *everything*—it seems unlikely that lorem-ipsum was interfering with expand-region—but I did not want to take the time to go package by package, mode by mode, and buffer by buffer to track down the exact problem(s).  Plus it was a good exercise in making sure that I could bootstrap a new account for my Emacs setup!)

**Comment by Trevoke** [comment]
I've blown away my entire `~/.emacs.d/elpa` repo a few times (I use `use-package`) in the past few weeks, for various reasons, the problem has not gone away.

Also - I just got emacs 24.5 from emacsforosx, blew away the elpa dir, installed. **expand-region works fine that way.**

This is a problem with emacs 25, at least on OSX, at least compiled this way:  `--with-cocoa --with-gnutls --with-imagemagick --with-mailutils --with-modules`

**Comment by mkleehammer** [comment]
Recompiling everything worked *for a while*, but it eventually was broken again.  I don't know if I restarted emacs or what, but it is definitely related to 25.1 and recompiling doesn't always fix it.

**Comment by Trevoke** [comment]
@magnars I'm going through the list of changes in emacs 25 and here's a few incompatible lisp changes that *might* affect this?

(to see the whole list, search for "Incompatible Lisp Changes in Emacs 25.1" in https://www.masteringemacs.org/article/whats-new-in-emacs-25-1 )

1.  The new implementation of Subword mode affects word movement everywhere.
When Subword mode is turned on, 'forward-word', 'backward-word', and
everything that uses them will move by sub-words, effectively
overriding the buffer's syntax table.  Lisp programs that shouldn't be
affected by Subword mode should call the new functions
'forward-word-strictly' and 'backward-word-strictly' instead.

2. 'save-excursion' does not save&restore the mark any more.
Use 'save-mark-and-excursion' if you want the old behavior.

3. 'deactivate-mark' is now buffer-local.

That was all I could find.

**Comment by magnars** [comment]
Thanks for your detective work on this, @Trevoke. I will install the latest Emacs and have a look at this as soon as I can. That may be some days tho.

**Comment by alphapapa** [comment]
> 'save-excursion' does not save&restore the mark any more. Use 'save-mark-and-excursion' if you want the old behavior.

Is it just me or is this change going to break a *lot* of packages?  :(

**Comment by Trevoke** [comment]
I think I found it. Not sure about the best fix.

I went to `er--expand-region-1` and saw the `ignore-errors`. I switched it to `with-demoted-errors` since the docs pointed me in that direction, and I saw these errors: 

> Error: (error "No url here")
> Error: (error "No email here")
> if: Symbol’s value as variable is void: expand-region-smart-cursor
>  [2 times]

Granted, no idea about the first two, BUT! The last one gave me an idea. I checked and indeed that variable did not exist. So I quickly did `(setq expand-region-smart-cursor nil)` and PRESTO! Expand-region works as desired.

I don't exactly know how to fix this, unfortunately, because that is clearly set as a `defcustom`. It is worth noting that this does *not* show up as an option in the `customize` interface.

**Comment by magnars** [comment]
Hmm. That is weird. Smart-cursor is our newest feature. So it's the one you'll likely not have compiled before. There's also autoloading involved in this picture. Does `(require 'expand-region-custom)` fix it?

**Comment by Trevoke** [comment]
It didn't -- and what was weird was that this morning "the fix" wasn't working. I was confused.

Then I realized that when I tried "my fix", I had cloned the most recent version of the code and modified that instead of what I have installed. I use melpa-stable, which means I was on version 0.10.0. I promptly switched to regular melpa, which got me the latest code, and everything works for me now.

So ... Fixing this might be as simple as tagging a new release for melpa stable.
